Google.Common/src/Google.Common.Embeddable/Preconditions.cs

```diff
@@ -0,0 +1,73 @@
+        // - Methods that return null or "something that will throw an exception" to be chained with string interpolation
+        // - Methods accepting a string format and one or two arguments (to avoid any arrays being created at execution time,
+        //   but still allow for string formatting)
+        // - Conditional public/internal access, so that we can create a Google.Common assembly exposing all of this publicly, without
+        //   duplication.
+
+        /// <summary>
+        /// Checks that the given argument (to the calling method) is non-null.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="argument"></param>
+        /// <param name="paramName">The name of the parameter in the calling method.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="argument"/> is null</exception>
+        /// <returns><paramref name="argument"/> if it is not null</returns>
+        internal static T CheckNotNull<T>(this T argument, string paramName) where T : class
+        {
+            if (argument == null)
+            {
+                throw new ArgumentNullException(paramName);
+            }
+            return argument;
+        }
+
+        /// <summary>
+        /// Checks that the given argument value is valid.
+        /// </summary>
+        /// <remarks>
+        /// Note that the upper bound (<paramref name="maxExclusive"/>) is inclusive,
+        /// not exclusive. This is deliberate, to allow the specification of ranges which include
+        /// <see cref="Int32.MaxValue"/>.
+        /// </remarks>
+        /// <param name="argument">The value of the argument passed to the calling method.</param>
+        /// <param name="paramName">The name of the parameter in the calling method.</param>
+        /// <param name="minInclusive">The smallest valid value.</param>
+        /// <param name="maxInclusive">The largest valid value.</param>
+        /// <exception cref="ArgumentOutOfRangeException">The argument was outside the specified range.</exception>
+        internal static void CheckArgumentRange(int argument, string paramName, int minInclusive, int maxInclusive)
+        {
+            if (argument < minInclusive || argument > maxInclusive)
+            {
+                throw new ArgumentOutOfRangeException(paramName, $"Value {argument} should be in range [{minInclusive}, {maxInclusive}]");
+            }
+        }
+
+        /// <summary>
+        /// Checks that given condition is met, throwing an <see cref="InvalidOperationException"/> otherwise.
+        /// </summary>
+        /// <param name="condition">The condition to test.</param>
+        /// <param name="message">The message to include in the exception, if generated. This should not
+        /// use interpolation, as the interpolation would be performed regardless of whether or not an exception is thrown.</param>
+        internal static void CheckState(bool condition, string message)
```

Comment:
But the array creation would occur anyway, which would be bad. I can add a couple of generic overloads to avoid any allocation - but I suggest we only go down that route when we find we need them.